### DISCUSSION TITLE
Beginner  here - import question

### DISCUSSION THREAD

**Discussion by jeff-breece** [juftin/camply#345] [Q&A] [answered]
- Using  VS  Code
- Following the basic OOP example in the docs
- Have a Venv running with the  Python 3.10.12 runtime
- Getting this error 'Import "camply.containers" could not be resolved'
![Screenshot from 2024-05-01 12-27-25](https://github.com/juftin/camply/assets/29708590/6d021d2f-bbfb-4209-b2db-b0686fe42810)
 
- **Question**,  how do I reference the assembly as I have the application installed, do I clone the repo to the lib folder or do I do something with the file path to the local install?

**Comment by juftin** [comment, marked as answer]
Hey @jeff-breece, I can see you're installing camply with [pipx](https://github.com/pypa/pipx). This is a utility meant to install and run Python applications in isolated virtual environments. 

Your use case however is to use camply as a library, and not a command line application. In this case you would want to `pip` install camply into your project's virtual environment. 

<details><summary>Generative AI Summary</summary>
<p>

> Explain the difference between Python applications that you would install with pipx and python libraries you would install with pip

Python, as a versatile programming language, offers a wide range of tools and libraries that cater to various needs, from web development and data analysis to automation and scientific computing. The Python ecosystem provides different package management tools to streamline the installation and management of these resources, with `pip` and `pipx` being two notable examples. Understanding the distinction between these tools and their appropriate use cases is crucial for effective Python development.

### pip

`pip` is the Python package installer that's been the standard for installing Python packages from the Python Package Index (PyPI) and other package indexes. It is used to manage (install, update, and remove) Python libraries within a particular Python environment. Libraries installed with `pip` are typically meant to be imported and used within your Python scripts or applications. For instance, if you're developing a Python project that requires requests for making HTTP requests, you would use `pip` to install it:

```sh
pip install requests
```

These libraries are dependencies of your project, meaning your code depends on them to run. `pip` is best used in conjunction with virtual environments (created using `venv` or `virtualenv`) to manage project-specific dependencies and avoid conflicts between different projects.

### pipx

`pipx` is a tool for installing and running Python applications in isolated environments. While `pip` is for libraries, `pipx` is designed for Python-based command-line tools and applications that are meant to be run in the terminal or command prompt as standalone applications, not imported into a Python script. Each application installed with `pipx` gets its own virtual environment, ensuring its dependencies do not interfere with those of other applications or globally installed Python packages.

For example, if you want to install a Python-based command-line tool like `black` (a code formatter), you would use `pipx`:

```sh
pipx install black
```

This approach ensures that `black` and its specific dependencies are isolated from other Python environments on your system. It's particularly useful for tools that you intend to use across multiple projects or Python environments, as it avoids version conflicts and keeps your system tidy.

### Key Differences

- **Scope**: `pip` is used for installing Python libraries that your projects depend on. `pipx` is used for installing Python applications that are meant to be run as standalone command-line tools.
- **Environment Isolation**: While both can be used in conjunction with virtual environments, `pipx` automatically isolates each installed application in its own environment by default.
- **Use Case**: Use `pip` when developing Python projects that need specific libraries. Use `pipx` when you want to install and run Python-based applications or tools that are not project-specific but are intended for general use across multiple projects or environments.

In summary, the choice between `pip` and `pipx` depends on what you're trying to achieve: `pip` for library management within projects, and `pipx` for installing and managing standalone Python applications.

</p>
</details>
